google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/RpcMethodMapping.java

```diff
@@ -0,0 +1,199 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage.conformance.retry;
+
+import static com.google.common.collect.Sets.newHashSet;
+import static java.util.Objects.requireNonNull;
+import static org.junit.Assert.fail;
+
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.conformance.retry.CtxFunctions.ResourceSetup;
+import com.google.cloud.storage.conformance.retry.Functions.CtxFunction;
+import com.google.common.base.Preconditions;
+import com.google.errorprone.annotations.Immutable;
+import java.util.HashSet;
+import java.util.function.Predicate;
+import org.junit.AssumptionViolatedException;
+
+/**
+ * Immutable class which represents a mapping between an {@link RpcMethod} and a method in the
+ * public {@code com.google.cloud.storage} API.
+ *
+ * <p>This class defines a semi-declarative why in which mappings can be declared independent of the
+ * actual environment & state necessary to actually invoke a method.
+ *
+ * @see ITRetryConformanceTest#test()
+ * @see RpcMethodMappings
+ */
+@Immutable
+final class RpcMethodMapping {
+
+  private final int mappingId;
+  private final RpcMethod method;
+  private final Predicate<TestRetryConformance> applicable;
+  private final CtxFunction setup;
+  private final CtxFunction test;
+  private final CtxFunction tearDown;
+
+  RpcMethodMapping(
+      int mappingId,
+      RpcMethod method,
+      Predicate<TestRetryConformance> applicable,
+      CtxFunction setup,
+      CtxFunction test,
+      CtxFunction tearDown) {
+    this.mappingId = mappingId;
+    this.method = method;
+    this.applicable = applicable;
+    this.setup = setup;
+    this.test = test;
+    this.tearDown = tearDown;
+  }
+
+  public int getMappingId() {
+    return mappingId;
+  }
+
+  public RpcMethod getMethod() {
+    return method;
+  }
+
+  public Predicate<TestRetryConformance> getApplicable() {
+    return applicable;
+  }
+
+  public CtxFunction getSetup() {
+    return setup;
+  }
+
+  public CtxFunction getTest() {
+    return (ctx, c) -> {
+      if (c.isExpectSuccess()) {
+        return test.apply(ctx, c);
+      } else {
+        try {
+          test.apply(ctx, c);
+          fail("expected failure, but succeeded");
+        } catch (StorageException e) {
+          // We expect an exception to be thrown by mapping and test retry conformance config
+          // Verify that the exception we received is actually what we expect.
+          boolean matchExpectedCode = false;
+          int code = e.getCode();
+          HashSet<String> instructions = newHashSet(c.getInstruction().getInstructionsList());
+          if (instructions.contains("return-503") && code == 503) {
+            matchExpectedCode = true;
+          }
+          if (instructions.contains("return-400") && code == 400) {
```

Comment:
Does this mean failures are valid even if they happen out of order?

Reply:
This code will only ever receive the last failure to occur in the code.

So in the case that we have a test which is expected to fail upon the first 400 returned by the server, this code will receive that 400.

In the case of something like `instructions: ["return-503", "return-503", "return-503", "return-503"]` this code doesn't know that the 503 it's receiving is from index 2 because the first two where expected to be handled by the retry logic, whereas the index 2 values would be due to retry exhaustion.

Reply:
Thanks for clarifying!